packages/firebase_vertexai/firebase_vertexai/example/lib/pages/function_calling_page.dart

```diff
@@ -0,0 +1,186 @@
+      'location': Schema.object(
+        description: 'The name of the city and its state for which to get '
+            'the weather. Only cities in the USA are supported.',
+        properties: {
+          'city': Schema.string(
+            description: 'The city of the location.',
+          ),
+          'state': Schema.string(
+            description: 'The state of the location.',
+          ),
+        },
+      ),
+      'date': Schema.string(
+        description: 'The date for which to get the weather. '
+            'Date must be in the format: YYYY-MM-DD.',
+      ),
+    },
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(widget.title),
+      ),
+      body: Padding(
+        padding: const EdgeInsets.all(8),
+        child: Column(
+          mainAxisAlignment: MainAxisAlignment.center,
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: [
+            Expanded(
+              child: ListView.builder(
+                itemBuilder: (context, idx) {
+                  return MessageWidget(
+                    text: _messages[idx].text,
+                    isFromUser: _messages[idx].fromUser ?? false,
+                  );
+                },
+                itemCount: _messages.length,
+              ),
+            ),
+            Padding(
+              padding: const EdgeInsets.symmetric(
+                vertical: 25,
+                horizontal: 15,
+              ),
+              child: Row(
+                children: [
+                  Expanded(
+                    child: ElevatedButton(
+                      onPressed: !_loading
+                          ? () async {
+                              await _testFunctionCalling();
+                            }
+                          : null,
+                      child: const Text('Test Function Calling'),
+                    ),
+                  ),
+                ],
+              ),
+            ),
+          ],
+        ),
+      ),
+    );
+  }
+
+  Future<void> _testFunctionCalling() async {
+    setState(() {
+      _loading = true;
+    });
+    final functionCallChat = _functionCallModel.startChat();
+    const prompt = 'What is the weather like in Boston on 10/02 this year?';
+
+    // Send the message to the generative model.
+    var response = await functionCallChat.sendMessage(
+      Content.text(prompt),
+    );
+
+    final functionCalls = response.functionCalls.toList();
+    // When the model response with a function call, invoke the function.
+    if (functionCalls.isNotEmpty) {
+      final functionCall = functionCalls.first;
+      if (functionCall.name == 'fetchWeather') {
+        Map<String, dynamic> location =
+            functionCall.args['location']! as Map<String, dynamic>;
+        var date = functionCall.args['date']! as String;
+        var city = location['city'] as String;
+        var state = location['state'] as String;
```

Comment:
I wonder if we could use a proper class with from/toJSON to handle this cleanly

Reply:
Like Location { String city, String state}? I feel we could, but would like to defer if we want to make more complicated function calling structure.